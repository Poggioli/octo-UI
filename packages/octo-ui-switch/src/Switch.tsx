import { forwardRef } from "@octo-ui/core";
import { StyledSwitchRoot, StyledSwitchThumb } from "./styles";
import { SwitchProps, SwitchThumbProps } from "./types";

const Switch = forwardRef<typeof StyledSwitchRoot, SwitchProps>(
  ({ ...props }, forwardedRef) => {
    return <StyledSwitchRoot {...props} ref={forwardedRef} />;
  }
);
Switch.toString = () => `.${StyledSwitchRoot.className}`;

// ========================================================================= //

const SwitchThumb = forwardRef<typeof StyledSwitchThumb, SwitchThumbProps>(
  ({ ...props }, forwardedRef) => {
    return <StyledSwitchThumb {...props} ref={forwardedRef} />;
  }
);
SwitchThumb.toString = () => `.${StyledSwitchThumb.className}`;

export { Switch, SwitchThumb };
export { StyledSwitchRoot, StyledSwitchThumb };

