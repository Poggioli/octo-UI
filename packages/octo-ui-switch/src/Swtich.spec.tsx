import { fireEvent, render } from "@testing-library/react";
import { Switch, SwitchContainer, SwitchLabel } from "./Switch";

describe("<Switch />", () => {
  it(`GIVEN a <Switch />
      WHEN render is called
      THEN should render a <Switch />`, () => {
    const rendered = render(<Switch />);
    expect(rendered.container).toMatchSnapshot();
  });

  it(`GIVEN a <Switch />, <SwitchLabel /> and <SwitchContainer />
      WHEN render it's enabled
      THEN should match to snapshot`, () => {
    const rendered = render(
      <SwitchContainer>
        <SwitchLabel htmlFor="switch-label">Switch Label</SwitchLabel>
        <Switch id="switch-label" />
      </SwitchContainer>
    );

    expect(rendered.container).toMatchSnapshot();
  });

  it(`GIVEN a <Switch />, <SwitchLabel /> and <SwitchContainer />
      WHEN render it's disabled
      THEN should match to snapshot`, () => {
    const rendered = render(
      <SwitchContainer>
        <SwitchLabel htmlFor="switch-label">Switch Label</SwitchLabel>
        <Switch id="switch-label" disabled />
      </SwitchContainer>
    );

    expect(rendered.container).toMatchSnapshot();
  });

  it(`GIVEN a <Switch />, <SwitchLabel /> and <SwitchContainer />
      WHEN render is called
      THEN should render a normaly`, () => {
    const rendered = render(
      <SwitchContainer>
        <SwitchLabel htmlFor="switch-label">Switch Label</SwitchLabel>
        <Switch id="switch-label" />
      </SwitchContainer>
    );

    const switchComponent = rendered.getByRole("switch");
    const switchLabel = rendered.getByText("Switch Label");

    expect(switchComponent).toBeVisible();
    expect(switchLabel).toBeVisible();
  });

  it(`GIVEN a <SwitchLabel />
      WHEN user click in label
      THEN should toggle switch state`, () => {
    const rendered = render(
      <SwitchContainer>
        <SwitchLabel htmlFor="switch-airplane">Switch Label</SwitchLabel>
        <Switch id="switch-airplane" />
      </SwitchContainer>
    );

    const switchLabel = rendered.getByText("Switch Label");
    fireEvent.click(switchLabel);

    const switchComponent = rendered.getByRole("switch", {
      checked: true,
    });

    expect(switchComponent).toBeVisible();
    expect(switchLabel).toBeVisible();
  });

  it(`GIVEN a <Switch />
      WHEN toString is called
      THEN return a string`, () => {
    expect(Switch.toString()).toBe(".octo-UI-c-PJLV");
  });

  it(`GIVEN a <SwitchContainer />
      WHEN toString is called
      THEN return a string`, () => {
    expect(SwitchContainer.toString()).toBe(".octo-UI-c-iQGaKg");
  });

  it(`GIVEN a <SwitchLabel />
      WHEN toString is called
      THEN return a string`, () => {
    expect(SwitchLabel.toString()).toBe(".octo-UI-c-gHWBzz");
  });
});
