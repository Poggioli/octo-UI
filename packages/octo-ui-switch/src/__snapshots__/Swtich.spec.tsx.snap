// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> GIVEN a <Switch />
      WHEN render is called
      THEN should render a <Switch /> 1`] = `
<div>
  <button
    aria-checked="false"
    class="octo-UI-c-liqGON octo-UI-c-liqGON-gYIKvC-color-error"
    data-state="unchecked"
    role="switch"
    type="button"
    value="on"
  >
    <span
      class="octo-UI-c-kBJywl"
      data-state="unchecked"
    />
  </button>
</div>
`;

exports[`<Switch /> GIVEN a <Switch />, <SwitchLabel /> and <SwitchContainer />
      WHEN render it's disabled
      THEN should match to snapshot 1`] = `
<div>
  <div
    class="octo-UI-c-iQGaKg"
  >
    <label
      class="octo-UI-c-gHWBzz octo-UI-c-kwAZbD octo-UI-c-gWerhV octo-UI-c-gWerhV-iMVRpC-size-3 octo-UI-c-gWerhV-iiZhlqR-css"
      for="switch-label"
    >
      Switch Label
    </label>
    <button
      aria-checked="false"
      class="octo-UI-c-liqGON octo-UI-c-liqGON-gYIKvC-color-error"
      data-disabled=""
      data-state="unchecked"
      disabled=""
      id="switch-label"
      role="switch"
      type="button"
      value="on"
    >
      <span
        class="octo-UI-c-kBJywl"
        data-disabled=""
        data-state="unchecked"
      />
    </button>
  </div>
</div>
`;

exports[`<Switch /> GIVEN a <Switch />, <SwitchLabel /> and <SwitchContainer />
      WHEN render it's enabled
      THEN should match to snapshot 1`] = `
<div>
  <div
    class="octo-UI-c-iQGaKg"
  >
    <label
      class="octo-UI-c-gHWBzz octo-UI-c-kwAZbD octo-UI-c-gWerhV octo-UI-c-gWerhV-iMVRpC-size-3 octo-UI-c-gWerhV-iisyIHw-css"
      for="switch-label"
    >
      Switch Label
    </label>
    <button
      aria-checked="false"
      class="octo-UI-c-liqGON octo-UI-c-liqGON-gYIKvC-color-error"
      data-state="unchecked"
      id="switch-label"
      role="switch"
      type="button"
      value="on"
    >
      <span
        class="octo-UI-c-kBJywl"
        data-state="unchecked"
      />
    </button>
  </div>
</div>
`;
